cmd/configurator/main.go

```diff
@@ -0,0 +1,317 @@
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"golang.org/x/sys/unix"
+
+	"google.golang.org/grpc"
+
+	"github.com/sylabs/slurm-operator/pkg/workload/api"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+)
+
+var (
+	redBoxSock = flag.String("sock", "", "path to red-box socket")
+
+	serviceAccount = os.Getenv("SERVICE_ACCOUNT")
+	kubeletImage   = os.Getenv("KUBELET_IMAGE")
+	hostNodeName   = os.Getenv("HOST_NAME")
+	namespace      = os.Getenv("NAMESPACE")
+)
+
+func main() {
+	flag.Parse()
+
+	if namespace == "" {
+		namespace = "default"
+	}
+
+	// getting k8s config.
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Fatalf("can't fetch cluster config %s", err)
+	}
+
+	// corev1 client set is required to create collecting results pod.
+	coreC, err := corev1.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("can't create core client %s", err)
+	}
+
+	conn, err := grpc.Dial("unix://"+*redBoxSock, grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("can't connect to %s %s", *redBoxSock, err)
+	}
+	slurmC := api.NewWorkloadManagerClient(conn)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go watchPartitions(ctx, wg, slurmC, coreC)
+
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, unix.SIGINT, unix.SIGTERM, unix.SIGQUIT)
+
+	log.Printf("Got signal %s", <-sig)
+	cancel()
+
+	wg.Wait()
+
+	log.Println("Configurator is finished")
+}
+
+func watchPartitions(ctx context.Context, wg *sync.WaitGroup, slurmClient api.WorkloadManagerClient, k8sClient *corev1.CoreV1Client) {
+	defer wg.Done()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.Tick(1 * time.Minute):
+			// getting SLURM partitions
+			partitionsResp, err := slurmClient.Partitions(context.Background(), &api.PartitionsRequest{})
+			if err != nil {
+				log.Printf("Can't get partitions %s", err)
+				continue
+			}
+
+			// gettings k8s nodes
+			nodes, err := k8sClient.Nodes().List(metav1.ListOptions{
+				LabelSelector: "type=virtual-kubelet",
+			})
+			if err != nil {
+				log.Printf("Can't get pods %s", err)
+				continue
+			}
+			// extract partition names from k8s nodes
+			nNames := partitionNames(nodes.Items)
+
+			// check which partitions are not yet represented in k8s
+			partitionToCreate := notIn(partitionsResp.Partition, nNames)
+			// creating pods for that partitions
+			if err := createNodeForPartitions(k8sClient, partitionToCreate); err != nil {
+				log.Printf("Can't create partitions  %s", err)
+			}
+
+			// some partitions can be deleted from SLURM, so we need to delete pods
+			// which are represent those deleted partitions
+			nodesToDelete := notIn(nNames, partitionsResp.Partition)
+			if err := deleteControllingPod(k8sClient, nodesToDelete); err != nil {
```

Comment:
Maybe we should start this switch case by removing orphaned pods and then trying to create new? Otherwise continue on errors above may make this step unreachable.

Reply:
but there is no continue on the create error above, so this step will be executed even if crate failed

Reply:
https://github.com/sylabs/slurm-operator/pull/141/files#diff-609a2676e042a2d7381aa11d0cc586f4R101

Reply:
nevermind, those are nodes